### PR TITLE
Bump go toolchain version to address CVE-2023-45288 for release-1.3

### PR DESCRIPTION
**Pull request by henrybear327**
Changes:
- Bump toolchain version to `1.21.9` due to CVE-2023-45288
- run `go mod tidy`

Address linter changes (align with the main branch):
- remove `rand.Seed(s)`
- use `rand.Read` from `"crypto/rand"`
- add `//nolint:all` for `(*reflect.SliceHeader)(slice)` -> will be fixed in a new PR

Reference:
- https://github.com/etcd-io/etcd/issues/17703

### CODE REVIEW

**Comment by henrybear327** [comment]
@ahrtr it's a huge bump from `1.17.13` to `1.21.9`.

**Comment by jmhbnz** [comment]
> @ahrtr it's a huge bump from `1.17.13` to `1.21.9`.

I believe the reference to `release-1.3` was in relation to `etcd-io/bbolt` repo not `etcd-io/etcd` repo.

**Comment by ivanvc** [comment]
Cross linking etcd-io/etcd#17703

**Comment by ivanvc** [comment]
Should we adopt the main branch's `.go-version` approach to defining the Go version to avoid mismatches in different files? What do you think @ahrtr?

**Comment by henrybear327** [comment]
> > Should we adopt the main branch's `.go-version` approach to defining the Go version to avoid mismatches in different files? What do you think @ahrtr?
> 
> Sounds good to me. Thanks

Thanks @ivanvc and @ahrtr, I will implement it.

**Comment by henrybear327** [comment]
> `.go-version` approach

@ivanvc @ahrtr, I have made my attempt at the `.go-version` approach!

File: cmd/bbolt/main_test.go [hunk 1]
```diff
@@ -304,12 +303,6 @@ func NewMain() *Main {
 }
 
 func TestCompactCommand_Run(t *testing.T) {
-	var s int64
-	if err := binary.Read(crypto.Reader, binary.BigEndian, &s); err != nil {
-		t.Fatal(err)
-	}
-	rand.Seed(s)
```
Comment (ahrtr) [inline cmd/bbolt/main_test.go:311]: Please raise a PR to remove them from `main` as well, thanks.

File: unsafe.go [hunk 2]
```diff
@@ -32,7 +32,7 @@ func unsafeByteSlice(base unsafe.Pointer, offset uintptr, i, j int) []byte {
 // manipulation of reflect.SliceHeader to prevent misuse, namely, converting
 // from reflect.SliceHeader to a Go slice type.
 func unsafeSlice(slice, data unsafe.Pointer, len int) {
-	s := (*reflect.SliceHeader)(slice)
+	s := (*reflect.SliceHeader)(slice) //nolint:staticcheck
```
Comment (ahrtr) [inline unsafe.go:35]: I think we need to backport https://github.com/etcd-io/bbolt/pull/566 to release-1.3.
- Remove `UnsafeSlice`, and use `unsafe.Slice`

It's OK to fix it in a followup

File: cmd/bbolt/main_test.go (hunk 1 above)
Comment (henrybear327) [reply cmd/bbolt/main_test.go:311]: Done according to request!

File: unsafe.go (hunk 2 above)
Comment (henrybear327) [reply unsafe.go:35]: Will do the porting in another PR!

**Comment by henrybear327** [comment]
After second thought, I bumped the go toolchain to `1.22.2`, as we were at `1.17.x` and there isn't a reason for us to just bump the version to `1.21.x`.

**Comment by ivanvc** [comment]
Please refer to https://github.com/etcd-io/etcd/issues/17348#issuecomment-1980562728. Unless @ahrtr thinks otherwise, I think it would be better to stay on 1.21 for the release-1.3 branch.

**Comment by henrybear327** [comment]
> Please refer to [etcd-io/etcd#17348 (comment)](https://github.com/etcd-io/etcd/issues/17348#issuecomment-1980562728). Unless @ahrtr thinks otherwise, I think it would be better to stay on 1.21 for the release-1.3 branch.

@ivanvc, thanks for the pointer. I will rework the PR.

**Comment by henrybear327** [comment]
> Please refer to [etcd-io/etcd#17348 (comment)](https://github.com/etcd-io/etcd/issues/17348#issuecomment-1980562728). Unless @ahrtr thinks otherwise, I think it would be better to stay on 1.21 for the release-1.3 branch.

@ahrtr which go toolchain version do we want to go with on `release-1.3` :) `1.21.9` or `1.22.2`

**Comment by henrybear327** [comment]
Rebased.

**Comment by henrybear327** [comment]
@ahrtr rebased and using go version 1.21.9!

**Review by ahrtr — APPROVED** [review]

**Review by ivanvc — APPROVED** [review]
Thanks @henrybear327

File: unsafe.go (hunk 2 above)
Comment (henrybear327) [reply unsafe.go:35]: Porting is attempted at https://github.com/etcd-io/bbolt/pull/717

**Comment by serathius** [comment]
We should avoid bumping go version in `go.mod` on stable releases as it means we drop support of older Go version which is a breaking change. 

Instead of bumping go version in `go.mod` we should add a separate CI job that tests multiple go version for compatibility.

**Comment by ahrtr** [comment]
> We should avoid bumping go version in `go.mod` on stable releases as it means we drop support of older Go version which is a breaking change.
> 
> Instead of bumping go version in `go.mod` we should add a separate CI job that tests multiple go version for compatibility.

https://github.com/etcd-io/bbolt/pull/747#issuecomment-2095764843

**Pull request merged**